### ISSUE TITLE
`tuple[*Ts]` subtype with extra type parameter not gradually assignable to itself

### ISSUE THREAD

**Issue by jorenham** [python/mypy#19908]
This demonstrates the issue:

```py
from typing import Any

class Array[DataT, *ShapeTs](tuple[*ShapeTs]): ...

type Float2D = Array[float, *tuple[int, int]]  # or Array[float, int, int]
type FloatND = Array[float, *tuple[Any, ...]]

def from_2d(a: Float2D) -> FloatND:
    return a  # mypy: ✅, pyright: ✅

def into_2d(a: FloatND) -> Float2D:
    return a  # mypy: ❌, pyright: ✅
```

https://mypy-play.net/?mypy=latest&python=3.13&flags=strict&gist=1b95c702df6de35fa3f61aa7b5dfef16

<details><summary>
Without the additional <code>DataT</code> typar, the false positive disappears</summary>
<p>

```py
from typing import Any

class Array[*ShapeTs](tuple[*ShapeTs]): ...

type Float2D = Array[*tuple[int, int]]  # or Array[int, int]
type FloatND = Array[*tuple[Any, ...]]

def from_2d(a: Float2D) -> FloatND:
    return a  # mypy: ✅, pyright: ✅

def into_2d(a: FloatND) -> Float2D:
    return a  # mypy: ✅, pyright: ✅
```

https://mypy-play.net/?mypy=latest&python=3.13&flags=strict&gist=be0dfedeb0e7ed14da359cd52d2901e7

</p>
</details> 

<details><summary>
The issue also disappears when tuple is no longer a base class</summary>
<p>

```py
from typing import Any

class Array[float, *ShapeTs]: ...

type Float2D = Array[float, *tuple[int, int]]  # or Array[float, int, int]
type FloatND = Array[float, *tuple[Any, ...]]

def from_2d(a: Float2D) -> FloatND:
    return a  # mypy: ✅, pyright: ✅

def into_2d(a: FloatND) -> Float2D:
    return a  # mypy: ✅, pyright: ✅
```

https://mypy-play.net/?mypy=latest&python=3.13&flags=strict&gist=a1439e7c3037239bc4663b981534fb17

</p>
</details> 

Potentially related issues:
- #18665
- #19106
- #19109
- #19110
- #19860 
- #19858
